### DISCUSSION TITLE
Using parameters with filters

### DISCUSSION THREAD

**Discussion by Elitek6** [thunderclient/thunder-client-support#1287] [Q&A] [answered]
In our project, we have to create IDs in the following way: 

`A constant prefix + a random string`

Generating the random string and attaching a prefix to it are easy, but our problem is that for every request, we have a different prefix. We have almost around 70 such requests. 

In Postman, we would save that prefix on the request level. Since this is not possible in Thunder Client, we are looking for a workaround and would appreciate any help. 

So far, the only workable option we have found is making a filter per request in our script file, and calling this filter in the appropriate request. Which means over 70 filters just to have different prefixes per request. Here is a simplified example of our script.

![image](https://github.com/rangav/thunder-client-support/assets/142809302/5ab502d0-88b7-4057-aa06-e5aceee26459)

What would make our lives easier is if we could use **parameters with filters**. Is this a feature you could consider adding? 

Or is there perhaps a different solution we have not considered? Thank you for your help. 

**Edit**: 
To clarify what I mean, here is a screenshot. I would like to be able to give parameters to filters. 

![image](https://github.com/rangav/thunder-client-support/assets/142809302/f9929837-ce84-473f-8d04-c82433597674)

**Comment by rangav** [comment]
Yes you can use paramater with filters, its already supported. 

But don't use the Pre Request Script feature.

Use the filters in Env Variable option.
https://github.com/rangav/thunder-client-support/blob/master/docs/filters.md#filters-in-env-var

Let me know if this is useful or not?

**Reply by Elitek6** [reply]
Hi Rangav, 

I'm sorry, I don't understand. 

Perhaps my use of the word "parameter" was confusing. I meant it as a parameter of a function, not a parameter of a request. My apologies if so. 

Let's say we have:

Request01
Request02
...
Request70

And every request has a json body and within that body, in specific places, it needs an ID with a prefix unique to it. 

How can I use filters in Environment variables here? Are you proposing that I save all 70 prefixes in the Environment? How would I select a specific one for, say, Request15?

**Reply by rangav** [reply]
Ok Will provide an example, give me few minutes.

**Comment by rangav** [comment, marked as answer]
Example Below

<img width="1101" alt="Screenshot 2023-08-21 at 15 58 40" src="https://github.com/rangav/thunder-client-support/assets/8637550/14285d75-554e-401c-8237-6ef82a591b91">



Filter Code:
```js
function randonId(input, prefix) {
    let randonVar = prefix + Math.random().toString(36).substring(7);
    console.log("randonVar", prefix, randonVar);
    tc.setVar("completeID", randonVar);
    return randonVar;
}

module.exports = [randonId]
```

Update the filter code as needed for your use-case

**Reply by Elitek6** [reply]
Thank you, this works. :)

**Reply by rangav** [reply]
Thanks for confirmation :)

**Comment by mrmaramreddy** [comment]
@rangav @Elitek6 

I am trying to implement exactly same thing listed in this topic and but not able to achieve it. Not sure if I am doing it wrong or it is broken with the version I am using.

**Steps to Reproduce:**
Create an external file with below content and attach it to the collection.
```javascript
function decrypt(stringToDecrypt) {
  console.log("String to decrypt = " + stringToDecrypt) //stringToDecrypt is null when called from HTTP header using {{@|decrypt("teststringtodecrypt")}}
  // decryption logic goes here and assign the result to 'decryptedValue' variable and return this variable
  return decryptedValue;
}

module.exports = [decrypt]
```

From HTTP headers (in the request), I am using below header (in raw format)
```
secret: {{@|decrypt("teststringtodecrypt")}}
```
This is invoking the "decrypt" function/filter in the external js file as expected but the value (in this case "teststringtodecrypt") is not passed to "decrypt" function/filter.

Am I doing something wrong??

**Platform:**
     OS: Mac
     vscode version: 1.83.1
     extension version: v2.14.0
**Are you using the free version/paid version/trial:**
     Free version

**Reply by rangav** [reply]
Hi @mrmaramreddy 

when you use `{{@|` you are not passing a variable to the filter function, then `stringToDecrypt` parameter will be `undefined`

change to `{{variableName | decrypt("teststringtodecrypt")}}` variableName will be name of variable where secret value is saved

**Comment by mrmaramreddy** [comment]
Thanks @rangav for the response. I am trying to pass hardcoded text (in this example) ```teststringtodecrypt``` as a parameter value to ```decrypt``` filter function, not a variable and trying to send the returned value from ```decrypt``` filter function as a ```secret``` header value in the http request. Are you saying I need to use variables (local or global or env) to pass the data between HTTP Headers and filter function?

I tried below with your suggestion, still see ```stringToDecrypt``` is null. I might be doing something wrong still. Appreciate any further help.

**external decrypt.js file**
```javascript
function decrypt(stringToDecrypt) {
  console.log("String to decrypt = " + stringToDecrypt) //stringToDecrypt is null when called from HTTP header using {{varTemp|decrypt("teststringtodecrypt")}}
  // decryption logic goes here and assign the result to 'decryptedValue' variable and return this variable
  tc.setVar("varTemp",decryptedValue,"global"); //Added this line to save decryptedValue in varTemp global variable 
  return decryptedValue; //if variables to be used, this return seems not required.
}

module.exports = [decrypt]
```

**Updated HTTP Header**
<img width="646" alt="image" src="https://github.com/rangav/thunder-client-support/assets/17325364/75171154-d305-451e-bd03-29eaba1ccbdb">

**Reply by rangav** [reply]
Ok got it 

change function to below format it will work in both cases.
- `function decrypt(input, stringToDecrypt)`


First Parameter will be variable value and 2nd parameter will be the value you pass in function.

you can use inline scripting feature if the filters is confusing.

**Comment by mrmaramreddy** [comment]
@rangav  Thank you for the help. Adding first argument ```input``` to the filter function did the trick. 

Regarding your ```inline scripting``` suggestion - did not find a way to reuse one function across multiple collections in inline scripting. I am trying to write common utils at one place and reuse them across multiple collections/requests which is possible with external script files. I used to partially achieve this in postman with below approach, may be similar thing exists in Thunder Client, did not find one yet.

Define the utils like below at collection level 

```javascript
//Common functions
utils = {
    function1: function(arg1){
     // function1 logic goes here
     return "function1 result"
    },
    function12: function(arg1, arg2){
     // function2 logic goes here
     return "function2 result"
    }
}
```

then use ```utils.function1("test")``` or ```utils.function2("test1", "test2")``` to access theses functions across different API requests inline scripting.

**Reply by rangav** [reply]
Thanks for confirmation,

If you want reusable functions then filters is the right way.

**Comment by rangav** [comment]
We launched a new feature - import functions from js files

https://github.com/rangav/thunder-client-support/releases/tag/v2.16.0
